core/src/local/persistent_state.rs

```diff
@@ -0,0 +1,891 @@
+use std::borrow::Cow;
+use std::rc::Rc;
+
+use bincode;
+use rocksdb::{self, SliceTransform, WriteBatch, DB};
+
+use ::*;
+use data::SizeOf;
+
+// Uniquely identifies an index for each base (the numeral index in PersistentState.indices).
+type IndexID = u32;
+
+// Incremented on each PersistentState initialization so that IndexSeq
+// can be used to create unique identifiers for rows.
+type IndexEpoch = u64;
+
+// Monotonically increasing sequence number since last IndexEpoch used to uniquely identify a row.
+type IndexSeq = u64;
+
+// RocksDB key used for storing meta information (like indices).
+const META_KEY: &'static [u8] = b"meta";
+
+struct PersistentIndex {
+    columns: Vec<usize>,
+    seq: IndexSeq,
+}
+
+// Store index information in RocksDB to avoid rebuilding indices on recovery.
+#[derive(Default, Serialize, Deserialize)]
+struct PersistentMeta {
+    indices: Vec<Vec<usize>>,
+    epoch: IndexEpoch,
+}
+
+impl PersistentIndex {
+    fn new(columns: Vec<usize>) -> Self {
+        Self { columns, seq: 0 }
+    }
+}
+
+/// PersistentState stores data in SQlite.
+pub struct PersistentState {
+    name: String,
+    db_opts: rocksdb::Options,
+    // We don't really want DB to be an option, but doing so lets us drop it manually in
+    // PersistenState's Drop by setting `self.db = None` - after which we can then discard the
+    // persisted files if we want to.
+    db: Option<DB>,
+    durability_mode: DurabilityMode,
+    indices: Vec<PersistentIndex>,
+    epoch: IndexEpoch,
+    primary_key: Option<Vec<usize>>,
+}
+
+impl SizeOf for PersistentState {
+    fn size_of(&self) -> u64 {
+        use std::mem::size_of;
+
+        size_of::<Self>() as u64
+    }
+
+    fn deep_size_of(&self) -> u64 {
+        self.size_of()
+    }
+}
+
+impl State for PersistentState {
+    fn process_records(&mut self, records: &mut Records, partial_tag: Option<Tag>) {
+        assert!(partial_tag.is_none(), "PersistentState can't be partial");
+        if records.len() == 0 {
+            return;
+        }
+
+        let mut batch = WriteBatch::default();
+        for r in records.iter() {
+            match *r {
+                Record::Positive(ref r) => {
+                    self.insert(&mut batch, r.clone());
+                }
+                Record::Negative(ref r) => {
+                    self.remove(&mut batch, r);
+                }
+                Record::BaseOperation(..) => unreachable!(),
+            }
+        }
+
+        // Sync the writes to RocksDB's WAL:
+        let mut opts = rocksdb::WriteOptions::default();
+        opts.set_sync(true);
```

Comment:
This seems really weird to me. Surely we should have already written to the WAL before we get to this point? Unless I'm misremembering GroupCommitQueue, acks are sent once the flush method there is called (which does nothing for RocksDB), and after that the writes are emitted into the graph (and processed by the bases). If we only write to the WAL here, then isn't there a window between when we ACK and when we write to the WAL here where we can crash and lose ACKed writes?

Reply:
#63 moved the write-ACKs to `process.rs` after `materialize` has been called, so `process_records` should finish before that happens - right?

Reply:
Ohhh, yeah, you're right. My bad.